### PR TITLE
Support bool type for tensors

### PR DESCRIPTION
**Pull request by alonre24**
We store tensor of type bool as a sequence of bits, where the minimal size is one byte (as UINT8).
Add validation that we do not overflow the size of the type of the tensor when setting new tensor.

### CODE REVIEW

File: src/redis_ai_objects/tensor.h [hunk 1]
```diff
@@ -31,6 +31,7 @@ static const char *RAI_DATATYPE_STR_INT32 = "INT32";
 static const char *RAI_DATATYPE_STR_INT64 = "INT64";
 static const char *RAI_DATATYPE_STR_UINT8 = "UINT8";
 static const char *RAI_DATATYPE_STR_UINT16 = "UINT16";
+static const char *RAI_DATATYPE_STR_BOOL = "BOOL";
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.h:34]:
```suggestion
static const char *RAI_DATATYPE_STR_BIT = "BIT";
```

File: src/redis_ai_objects/tensor.c [hunk 2]
```diff
@@ -55,10 +73,13 @@ DLDataType RAI_TensorDataTypeFromString(const char *typestr) {
             return (DLDataType){.code = kDLUInt, .bits = 16, .lanes = 1};
         }
     }
+    if (strcasecmp(typestr, "BOOL") == 0) {
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:76]:
```suggestion
    if (strcasecmp(typestr, "BIT") == 0) {
```

File: src/redis_ai_objects/tensor.c [hunk 3]
```diff
@@ -24,6 +25,23 @@
 
 extern RedisModuleType *RedisAI_TensorType;
 
+// Check if the given value is in the range of the tensor type.
+bool _ValOverflow(long long val, RAI_Tensor *t) {
+    DLDataType dtype = t->tensor.dl_tensor.dtype;
+    if (dtype.code == kDLInt) {
+        uint max_abs_val = (uint)1 << (uint)(dtype.bits - 1);
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:32]:
```suggestion
        long long max_abs_val = (uint)1 << (uint)(dtype.bits - 1) -1 ;
```

File: src/redis_ai_objects/tensor.c [hunk 4]
```diff
@@ -24,6 +25,23 @@
...
+        if (val >= (long long)max_abs_val || val <= (long long)-1 * max_abs_val) {
+            return true;
+        }
+    } else if (dtype.code == kDLUInt) {
+        uint max_val = (uint)1 << dtype.bits;
+        if (val >= max_val || val < 0) {
+            return true;
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:39]: please add a comment that we support up to 32 bits of unsigned int
also, please keep all the types in this functions as `long long`

File: src/redis_ai_objects/tensor.c [hunk 5]
```diff
@@ -92,6 +113,9 @@ int Tensor_DataTypeStr(DLDataType dtype, char *dtypestr) {
         } else if (dtype.bits == 16) {
             strcpy(dtypestr, RAI_DATATYPE_STR_UINT16);
             result = REDISMODULE_OK;
+        } else if (dtype.bits == 1) {
+            strcpy(dtypestr, RAI_DATATYPE_STR_BOOL);
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:117]:
```suggestion
            strcpy(dtypestr, RAI_DATATYPE_STR_BIT);
```

File: src/redis_ai_objects/tensor.c [hunk 6]
```diff
@@ -129,9 +153,10 @@ RAI_Tensor *RAI_TensorCreateWithDLDataType(DLDataType dtype, long long *dims, in
     DLDevice device = (DLDevice){.device_type = kDLCPU, .device_id = 0};
 
     // If we return an empty tensor, we initialize the data with zeros to avoid security
-    // issues. Otherwise, we only allocate without initializing (for better performance)
+    // issues. Otherwise, we only allocate without initializing (for better performance).
+    // Also, for boolean tensors we should initialize data with zeros.
     void *data;
-    if (empty) {
+    if (empty || dtype.bits == 1) {
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:159]: why?

File: src/redis_ai_objects/tensor.c [hunk 7]
```diff
@@ -414,6 +439,12 @@ int RAI_TensorSetValueFromLongLong(RAI_Tensor *t, long long i, long long val) {
         }
     } else if (dtype.code == kDLUInt) {
         switch (dtype.bits) {
+        case 1:
+            // If the val is 1, set the corresponding bit.
+            if (val % 2) {
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:444]:
```suggestion
            if (val ==1) {
```

**Review by DvirDukhan — CHANGES_REQUESTED** [review]
looks good
I think this PR should add the support for the appropriate backends also

File: src/redis_ai_objects/tensor.c [hunk 8]
```diff
@@ -24,6 +25,27 @@
 
 extern RedisModuleType *RedisAI_TensorType;
 
+// Check if the given value is in the range of the tensor type.
+bool _ValOverflow(long long val, RAI_Tensor *t) {
+    DLDataType dtype = t->tensor.dl_tensor.dtype;
+    if (dtype.code == kDLInt) {
+        unsigned long long max_abs_val = ((unsigned long long)1 << (uint)(dtype.bits - 1)) - 1;
+        if (val > max_abs_val || val <= -1 * (long long)max_abs_val) {
```
Comment (DvirDukhan) [inline src/redis_ai_objects/tensor.c:33]: `if (val > max_abs_val`
is signed and unsigned comparison, any edge cases?

**Review by lantiga — APPROVED** [review]
Looks good!

File: src/backends/tflite.c [hunk 9]
```diff
@@ -32,7 +32,7 @@ RAI_Model *RAI_ModelCreateTFLite(RAI_Backend backend, const char *devicestr, RAI
         dl_device = kDLCPU;
         break;
     case RAI_DEVICE_GPU:
-        dl_device = kDLGPU;
+        dl_device = kDLCUDA;
```
Comment (DvirDukhan) [inline src/backends/tflite.c:35]: this is a breaking change
either introduce a new rdb encoding version or undo this change.

File: src/backends/tensorflow.c [hunk 10]
```diff
@@ -23,38 +23,38 @@ TF_DataType RAI_GetTFDataTypeFromDL(DLDataType dtype) {
...
-            break;
         case 16:
             return TF_UINT16;
-            break;
+        default:
+            return 0;
+        }
+    } else if (dtype.code == kDLBool) {
+        switch (dtype.bits) {
+        case 1:
```
Comment (DvirDukhan) [inline src/backends/tensorflow.c:56]: 8?

File: src/backends/libtorch_c/torch_c.cpp [hunk 11]
```diff
@@ -68,7 +69,7 @@ static DLDevice getDLDevice(const at::Tensor &tensor, const int64_t &device_id)
     DLDevice device;
     device.device_id = device_id;
     if (tensor.is_cuda()) {
-        device.device_type = DLDeviceType::kDLGPU;
+        device.device_type = DLDeviceType::kDLCUDA;
```
Comment (DvirDukhan) [inline src/backends/libtorch_c/torch_c.cpp:72]: breaking change

**Review by DvirDukhan — CHANGES_REQUESTED** [review]

File: src/backends/libtflite_c/tflite_c.cpp [hunk 12]
```diff
@@ -60,7 +64,7 @@ static at::DeviceType getATenDeviceType(DLDeviceType device_type) {
   switch (device_type) {
     case DLDeviceType::kDLCPU:
       return at::DeviceType::CPU;
-    case DLDeviceType::kDLGPU:
+    case DLDeviceType::kDLCUDA:
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/backends/libtflite_c/tflite_c.cpp:67]: I guess so...

File: src/backends/torch.c [hunk 13]
```diff
@@ -51,7 +51,7 @@ RAI_Model *RAI_ModelCreateTorch(RAI_Backend backend, const char *devicestr, RAI_
         dl_device = kDLCPU;
         break;
     case RAI_DEVICE_GPU:
-        dl_device = kDLGPU;
+        dl_device = kDLCUDA;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/backends/torch.c:54]: This isn't a breaking change. dlpack renamed this in their last release: https://github.com/dmlc/dlpack/releases
but the numeric value has not changed.

File: src/backends/tensorflow.c (hunk 10 above)
Comment (alonre24) [reply src/backends/tensorflow.c:56]: right! thanks

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**